modules/effects/src/effect_creator.ts

```diff
@@ -42,30 +44,25 @@ type ObservableReturnType<T> = T extends false
  */
 export function createEffect<
   C extends EffectConfig,
-  T extends DispatchType<C>,
-  O extends ObservableReturnType<T>,
-  R extends O | ((...args: any[]) => O)
->(source: () => R, config?: Partial<C>): R {
+  DT extends DispatchType<C>,
+  OT extends ObservableType<DT, OT>,
+  R extends Observable<OT> | ((...args: any[]) => Observable<OT>)
+>(source: () => R, config?: Partial<C>): R & CreateEffectMetadata {
   const effect = source();
-  // Right now both createEffect and @Effect decorator set default values.
-  // Ideally that should only be done in one place that aggregates that info,
-  // for example in mergeEffects().
   const value: EffectConfig = {
-    dispatch: true,
-    resubscribeOnError: true,
+    ...DEFAULT_EFFECT_CONFIG,
     ...config, // Overrides any defaults if values are provided
   };
   Object.defineProperty(effect, CREATE_EFFECT_METADATA_KEY, {
     value,
   });
-  return effect;
+  return effect as typeof effect & CreateEffectMetadata;
 }
 
-export function getCreateEffectMetadata<T>(instance: T): EffectMetadata<T>[] {
-  const propertyNames = Object.getOwnPropertyNames(instance) as Extract<
-    keyof T,
-    string
-  >[];
+export function getCreateEffectMetadata<
+  T extends { [props in keyof T]: Object }
```

Comment:
What's the difference between the following?

```ts
T extends { [props in keyof T]: Object }
```

VS

```ts
T extends Object 
```

Reply:
That's a really good question, @timdeschryver.
We are accessing the property and checking `instance[propertyName].hasOwnProperty(...)`. With `T extends Object` there's no info about the what properties there are (and they could be `undefined` with TS 3.5 and thus not allowed to call `hasOwnProperty`).

At some point I put it behind the type guard, but that PR got way too big and without focus. I think I'll adjust it soon.

Reply:
Would the following cross the line? 

```
T extends {
    [props in keyof T]: { [CREATE_EFFECT_METADATA_KEY]: EffectMetadata<T> }
  }
```

Because, we're stilling casting to `any` to access the `metaData`:

```
const metaData = (instance[propertyName] as any)[
        CREATE_EFFECT_METADATA_KEY
];
```

Reply:
Got it, thanks for the info Alex!

Would the following cross the line? 

```
T extends {
    [props in keyof T]: { [CREATE_EFFECT_METADATA_KEY]: EffectMetadata<T> }
  }
```

Because, we're stilling casting to `any` to access the `metaData`:

```
const metaData = (instance[propertyName] as any)[
        CREATE_EFFECT_METADATA_KEY
];
```